### ISSUE TITLE
Snippet, Buffer Sources Not Being Loaded

### ISSUE THREAD

**Issue by Drew-Daniels** [Saghen/blink.cmp#132]
## Original:
Thanks for making this plugin. Seems like a speedier alternative to the previous `nvim-cmp` and `luasnip` setup I was using for awhile. 

### Issue
This plugin worked out of the box, and I only ran into snags after trying to port my `luasnip` snippets over to the VSCode format so I can use them with `blink.cmp`. 

But it looks like my local snippets aren't getting loaded, and when I try to adjust my configuration so that they are, I lose all the out-of-the-box functionality. I guess I'm not sure what the configuration needs to look like for this to happen for both `friendly-snippets` and custom snippets to work together.

### Context
I think my confusion stems from a couple things:

#### 1. `sources.providers` configuration confusion
It looks like the default `sources.providers` configuration is completely overridden if we pass anything to `sources.providers`

Maybe this is intentional (or it works differently), but it seems like a difficult way for users to configure things, especially when it's not super clear to me in the docs what these defaults are. 

Here's what I see when my config looks like this:
```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- no provider.sources configuration
})
```

<img width="588" alt="Test_filter(data)" src="https://github.com/user-attachments/assets/7165354f-4160-434a-92e8-9fca021809d8">

... But as soon as I configure anything in `providers.sources`, I lose all of those completions, and my custom snippets don't appear either:

```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- TODO: Need to figure out why custom snippets aren't being sourced
  sources = {
    providers = {
      { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3 },
    },
  },
})
```

<img width="572" alt="show_docunentation = C-k," src="https://github.com/user-attachments/assets/26122d8b-0a73-4549-862c-7be6538410f4">

#### 2. Unsure how to configure custom snippets in addition to other snippet sources, like `friendly-snippets`

Even when I have a `provider` like the following, and I specify the path to where my snippets are, they aren't being displayed as options for completion. Other snippets are - but not my custom ones:

```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  sources = {
    providers = {
      { "blink.cmp.sources.lsp", name = "LSP" },
      { "blink.cmp.sources.path", name = "Path", score_offset = 3 },
      { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3, opts = { search_paths = {vim.fn.stdpath("config") .. "/snippets" } } },
      { "blink.cmp.sources.buffer", name = "Buffer", fallback_for = { "LSP" } },
    },
  },
})
```

<img width="588" alt="Test_filter(data)" src="https://github.com/user-attachments/assets/7165354f-4160-434a-92e8-9fca021809d8">

### Abbr `nvim/plugins/init.lua`:

```lua
return require("lazy").setup({
  {
    "chrisgrieser/nvim-scissors",
    dependencies = { "nvim-telescope/telescope.nvim", "garymjr/nvim-snippets" },
  },
  {
    "saghen/blink.cmp",
    lazy = false, -- lazy loading handled internally
    -- optional: provides snippets for the snippet source
    dependencies = "rafamadriz/friendly-snippets",
    -- use a release tag to download pre-built binaries
    version = "v0.*",
    -- OR build from source, requires nightly: https://rust-lang.github.io/rustup/concepts/channels.html#working-with-nightly-rust
    -- build = 'cargo build --release',
    -- On musl libc based systems you need to add this flag
    -- build = 'RUSTFLAGS="-C target-feature=-crt-static" cargo build --release',
  },
})
```

`uname -a`:
```sh
Darwin drews-m1-mbp 23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6000 arm64
```

`nvim --version`:
```sh
❯ nvim --version
NVIM v0.10.2
Build type: Release
LuaJIT 2.1.1727870382
Run "nvim -V1 -v" for more info
```

`nvim/snippets/lua.json`:
- This is the snippet I am trying to have show up listed in my completions, but no luck yet:
```json
{
  "TODO": {
    "body": "-- TODO: ${1:TODO}",
    "prefix": "td"
  }
}
```

**Comment by Drew-Daniels** [comment]
I've been doing some more investigation into this and it looks like a couple things are happening on my end:
1. Snippets (`friendly-snippets` for now, gave up on trying to get custom snippets loaded for now) do not appear to be getting loaded as completion sources
2. Buffer completions also appear to not be usable either

### `lazy.nvim` configuration

```lua
return require("lazy").setup({
  {
    "saghen/blink.cmp",
    lazy = false, -- lazy loading handled internally
    -- optional: provides snippets for the snippet source
    dependencies = "rafamadriz/friendly-snippets",
    -- use a release tag to download pre-built binaries
    version = "v0.*",
    -- OR build from source, requires nightly: https://rust-lang.github.io/rustup/concepts/channels.html#working-with-nightly-rust
    -- build = 'cargo build --release',
    -- On musl libc based systems you need to add this flag
    -- build = 'RUSTFLAGS="-C target-feature=-crt-static" cargo build --release',
  },
})
```

### LSP Completions
LSP completions work fine without any issues, when I rely on the `sources` defaults, or when set `sources.providers` as I do here:
```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- TODO: Need to figure out why custom snippets aren't being sourced
  sources = {
    providers = {
      { "blink.cmp.sources.lsp", name = "LSP" },
      -- { "blink.cmp.sources.path", name = "Path", score_offset = 3 },
      -- { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3 },
      -- { "blink.cmp.sources.buffer", name = "Buffer", fallback_for = { "LSP" } },
    },
  },
})
```

<img width="335" alt="clang-format," src="https://github.com/user-attachments/assets/983edbc9-7215-420d-b393-0d6f582a8964">

### Path Completions

```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- TODO: Need to figure out why custom snippets aren't being sourced
  sources = {
    providers = {
      -- { "blink.cmp.sources.lsp", name = "LSP" },
      { "blink.cmp.sources.path", name = "Path", score_offset = 3 },
      -- { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3 },
      -- { "blink.cmp.sources.buffer", name = "Buffer", fallback_for = { "LSP" } },
    },
  },
})
```

<img width="561" alt="image" src="https://github.com/user-attachments/assets/626d6229-7fa1-41e9-b941-f0be7a62f651">

### Snippet Completions
I have listed `friendly-snippets` as a `blink.cmp` dependency, but am not seeing any of them suggested as a type triggers  of snippets like the one for `print`, which has a trigger of `p`:

<img width="357" alt="image" src="https://github.com/user-attachments/assets/552d1a11-82fe-4067-b1da-553f81a3e257">

```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- TODO: Need to figure out why custom snippets aren't being sourced
  sources = {
    providers = {
      -- { "blink.cmp.sources.lsp", name = "LSP" },
      -- { "blink.cmp.sources.path", name = "Path", score_offset = 3 },
      { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3 },
      -- { "blink.cmp.sources.buffer", name = "Buffer", fallback_for = { "LSP" } },
    },
  },
})
```

<img width="361" alt="image" src="https://github.com/user-attachments/assets/97adac98-537f-424b-9c07-75f4c3a4622b">

### Buffer Completions
As a type words that I know have been used in the buffer I am editing, I do not see them suggested, like `print` in the example below:

```lua
require("blink-cmp").setup({
  keymap = {
    scroll_documentation_up = "<C-u>",
    scroll_documentation_down = "<C-d>",
    select_next = "<C-n>",
    select_prev = "<C-p>",
    accept = "<C-e>",
    show = "<C-space>",
    hide = "<C-space>",
    show_documentation = "<C-k>",
    hide_documentation = "<C-k>",
    snippet_forward = "<C-j>",
    snippet_backward = "<C-h>",
  },
  nerd_font_variant = "mono",
  highlight = {
    use_nvim_cmp_as_default = true,
  },
  -- TODO: Need to figure out why custom snippets aren't being sourced
  sources = {
    providers = {
      -- { "blink.cmp.sources.lsp", name = "LSP" },
      -- { "blink.cmp.sources.path", name = "Path", score_offset = 3 },
      -- { "blink.cmp.sources.snippets", name = "Snippets", score_offset = -3 },
      { "blink.cmp.sources.buffer", name = "Buffer", fallback_for = { "LSP" } },
    },
  },
})
```

<img width="578" alt="image" src="https://github.com/user-attachments/assets/7f404350-12d9-469e-bd1f-e1e5b08ae3f3">

**Comment by Saghen** [comment]
> It looks like the default sources.providers configuration is completely overridden if we pass anything to sources.providers

> Maybe this is intentional (or it works differently), but it seems like a difficult way for users to configure things, especially when it's not super clear to me in the docs what these defaults are.

The defaults can be found in the README or in `config.lua`. Overriding the entire list when merging is standard behavior when configuring in neovim but I agree it's confusing. Just not sure what a good alternative would be

**Comment by Drew-Daniels** [comment]
> Buffer source isn't showing because you have `fallback_for = { "LSP" }` but no source named `LSP` for it to fallback for. The snippets I'm not sure and I'm not able to reproduce it. If you have the chance, you could debug the `register_snippets` function here
> 
> https://github.com/Saghen/blink.cmp/blob/213fd94de2ab83ff409e1fd240625959bf61624e/lua/blink/cmp/sources/snippets/scan.lua?plain=1#L4-L30

Oh, that makes sense. Thanks for pointing that out. I can confirm that removing `fallback_for = { 'LSP' }` fixed it.

I'll try to take a look at the `register_snippets` function later today

**Comment by Drew-Daniels** [comment]
> > It looks like the default sources.providers configuration is completely overridden if we pass anything to sources.providers
> 
> > Maybe this is intentional (or it works differently), but it seems like a difficult way for users to configure things, especially when it's not super clear to me in the docs what these defaults are.
> 
> The defaults can be found in the README or in `config.lua`. Overriding the entire list when merging is standard behavior when configuring in neovim but I agree it's confusing. Just not sure what a good alternative would be

I see, thanks for confirming how that works. I think I was mostly confused because of how the defaults are shown in the `README` - since there are two `providers` keys listed. Might be obvious to others that the second is the full default config, but I wasn't sure if the configurations in the first `providers` key were using the same `score_offsets`, `opts`, etc. as the second one.

<img width="903" alt="image" src="https://github.com/user-attachments/assets/548c8d5d-ebb9-401b-bfa2-b582da590e6e">

**Comment by Drew-Daniels** [comment]
Ok, figured out what the issue was finally. 

When I was working on migrating from `luasnip` and `nvim-cmp` to `blink.cmp`, I had created a bunch of empty `<language>.json` files in my `nvim/snippets` directory (`all.json`, `lua.json`, `ruby.json` etc.) and my `package.json` looked something like this (I've abbreviated it here):

```jsonc
{
  "name": "my-snippets",
  "contributes": {
    "snippets": [
      {
        "language": "all",
        "path": "./all.json" // empty
      },
      {
        "language": "lua",
        "path": "./lua.json"
      },

      {
        "language": "ruby", 
        "path": "./ruby.json" // empty
      }
    ]
  }
}
```

`nvim/snippets/lua.json`:

```json
{
  "TODO": {
    "body": "-- TODO: ${1:TODO}",
    "prefix": "td"
  }
}
```

Once I removed all language entries for all the languages that mapped to empty `.json` snippet files in my `package.json`, and removed all those blank `.json` files from `nvim/snippets`, I saw both `friendly-snippets` and custom snippets displayed as completion sources the next time I opened `neovim`:

<img width="362" alt="image" src="https://github.com/user-attachments/assets/53ed5ad5-cb5e-4fdb-930d-b1c481b2831c">

<br>

<img width="403" alt="image" src="https://github.com/user-attachments/assets/9a2e003c-7690-4001-ad12-cad03d9a3780">

I'm wondering if there should be some validation done by `blink.cmp` to alert users if they've have invalid snippet files, or add error handling that skips over these files, but either way I think my case was kind of a weird edge case since I don't expect many people to have blank `.json` files like I did in my `nvim/snippets` directory.

Anyways, thanks for the help. Pointing out the `register_snippets` function helped a ton in figuring out what the issue was, and I'm excited to start using this plugin!
